src/commcare_cloud/commands/deploy/commcare.py

```diff
@@ -32,19 +52,115 @@ def confirm_deploy(deploy_revs, diffs, args):
+    ):
+        return False
+
+    diff = _get_diff(environment, deploy_revs)
+    diff.print_deployer_diff()
+    if diff.deployed_commit_matches_latest_commit and not args.quiet:
+        _print_same_code_warning(deploy_revs['commcare'])
+    return ask(
+        'Are you sure you want to preindex and deploy to '
+        '{env}?'.format(env=environment.name), quiet=args.quiet)
+
+
+@memoized
+def _get_diff(environment, deploy_revs):
+    repo = get_github().get_repo('dimagi/commcare-hq') if github_auth_provided() else None
+
+    deployed_version = _get_deployed_version(environment)
+    latest_version = repo.get_commit(deploy_revs['commcare']).sha if repo else None
+
+    return DeployDiff(repo, deployed_version, latest_version, new_version_details=deploy_revs)
+
+
+def _confirm_translated(environment, quiet=False):
+    if datetime.now().isoweekday() != 3 or environment.meta_config.deploy_env != 'production':
+        return True
+    return ask(
+        "It's the weekly Wednesday deploy, did you update the translations "
+        "from transifex? Try running this handy script from the root of your "
+        "commcare-hq directory:\n./scripts/update-translations.sh\n",
+        quiet=quiet
+    )
+
+
+def _confirm_environment_time(environment, quiet=False):
+    window = environment.fab_settings_config.acceptable_maintenance_window
+    if window:
+        d = datetime.now(pytz.timezone(window['timezone']))
+        if window['hour_start'] <= d.hour < window['hour_end']:
+            return
+    else:
+        return
+
+    message = (
+        "Whoa there bud! You're deploying '%s' outside the configured maintenance window. "
+        "The current local time is %s.\n"
+        "ARE YOU DOING SOMETHING EXCEPTIONAL THAT WARRANTS THIS?"
+    ) % (environment.name, d.strftime("%-I:%M%p on %h. %d %Z"))
+    return ask(message, quiet=quiet)
+
+
+def _print_same_code_warning(code_branch):
+    if code_branch == 'master':
+        branch_specific_msg = "Perhaps you meant to merge a PR or specify a --set code_branch=<branch> ?"
+    elif code_branch == 'enterprise':
+        branch_specific_msg = (
+            "Have you tried rebuilding the enterprise branch (in HQ directory)? "
+            "./scripts/rebuildstaging --enterprise"
+        )
+    elif code_branch == 'autostaging':
+        branch_specific_msg = (
+            "Have you tried rebuilding the autostaging branch (in HQ directory)? "
+            "./scripts/rebuildstaging"
+        )
+    else:
+        branch_specific_msg = (
+            "Did you specify the correct branch using --set code_branch=<branch> ?"
+        )
+
+    message = (
+        "Whoa there bud! You're deploying {code_branch} which happens to be "
+        "the same code as was previously deployed to this environment.\n"
+        "{branch_specific_msg}\n"
+        "Is this intentional?"
+    ).format(code_branch=code_branch, branch_specific_msg=branch_specific_msg)
+    print(message)
```

Comment:
nit: we're on Python 3 now, so f-strings are fair game.
```suggestion
    print(
        f"Whoa there bud! You're deploying {code_branch} which happens to be "
        f"the same code as was previously deployed to this environment.\n"
        f"{branch_specific_msg}\n"
        f"Is this intentional?"
    )
```